### ISSUE TITLE
Allow Category to be a Dropdown

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#2368]
### Background information

- [x] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [x] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [x] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.3.0
- OSPOS git commit hash is: 
- PHP version is: 5.6.38
- MySQL or MariaDB version is: MySQL 5.7
- OS and version is: FreeBSD 11.2
- WebServer is: Apache 2.4
- Selected language is: English
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### New Feature
Description and category in the items table feel like attributes to me.  Is there a reason why we shouldn't migrate them to attributes?
#851 as least seems to be requesting this for category so they can make it a dropdown.  For me at least this is important because then I can make category a dropdown so my staff can stop messing up the data.

**Comment by jekkos** [comment]
I had the discussion with @daN4cat earlier on and at that time he did not agree to make category an attribute.  Beside this I do agree that for me it would also be more convenient to have a dropdown instead of free text for this field. But then of course you need another screen in which you can enter the values, which brings us back to attributes. 

So changing it as it is now to become a dropdown is not really an option either. But we could foresee some thing like a fixed 'system' type of attribute that is present by default and cannot be deleted.

**Comment by objecttothis** [comment]
@daN4cat perhaps there is a good reason to keep category and possibly also description the way they are. Maybe you could help me understand.

**Comment by daN4cat** [comment]
Going extreme every entity of a digital representation of an asset could be an attribute, if we want to be purist in the object oriented modelling.
However we struck a balance between what is the core part of the data model and what is additional.
Title is the name of the asset, Description gives more context and Category is a basic means to have a taxonomy for that asset (e.g. books, shoes, bread), etc.
We keep all of that in the same core data model that is then one entry in a table.
Attributes are any additional decoration, an enrichment to the data model or simply metadata (e.g. blue, english, handmade).
We could argue on data modelling and dimensional views, we don't have the two separate and we balanced a bit between the two to keep queries simple.

**Comment by daN4cat** [comment]
Having Category as free text makes easy to suggest or add a new one. More freedom and natural addition.

**Comment by objecttothis** [comment]
Ok, so is there anything we can do to prevent users from screwing the data up with Category?

**Comment by objecttothis** [comment]
I don't need it to be an attribute really. My goal is to stop the data from getting messed up.

**Comment by objecttothis** [comment]
What about keeping it in the items table but optionally allowing admins to make it a drop-down for those of us that want more data protection?

**Comment by daN4cat** [comment]
Fine with the dropbox as optional. I can see the benefits.

**Comment by jekkos** [comment]
We should make it less obvious to add a non existing or new value through the same widget. As @objecttothis most of the time you want to limit yourself to a fixed set of categories so you can use it for reporting purposes. I think @daN4cat you proposed some solution to this in the form of using a specific dropdown that still allows you to enter unseen values, but suggests what exists as a first option.

Would that suit your needs @objecttothis or do you need to disable users adding new categories completely? For example I see that bootstrap-select offers an option that combines both dropdown and free text in one control.

**Comment by objecttothis** [comment]
I'll take a look at bootstrap-select to see what it offers, but it doesn't seem to matter how many times we ask our staff to pay attention to what they are entering, they mess it up.  Part of the problem is what gets manually entered through the view and the other part of the problem is what gets imported through the CSV import feature.  With attributes when I implement the CSV/attribute import stuff I'm also going to add checks for dropdowns and not allow importing of fields that are not a dropdown option.  If we offer an optional dropdown for Category we should do the same kind of checks when importing (but only if Category is a dropdown) to prevent junk data from manually getting entered or imported.

This would be much easier if Category could be an attribute since attributes can be free text or dropdowns, but its needed (by me) enough that I'm willing to do the extra work to make an optional dropdown if we aren't making it an attribute.

**Comment by jekkos** [comment]
Ok I guess there are some advantages and disadvantages here.

Making it an attribute will require us to join in two more tables in most of the already existing reports, as they refer to category directly in many queres. I think it might certainly complicate the fallout of the initial change which seems rather easy.

So in terms of datamodel it might not be the best move. On the other hand using the bootstrap select might cost you some more time on the UI side, but you

**Comment by objecttothis** [comment]
@jekkos Sounds good.  You are right that there would certainly be fallout from making category an attribute.  We will go the route of keeping it as is and adding an optional feature to make it a dropdown.  Any preference on where in the database to store the dropdown options?  We could store it in the app_config table as a serialized array but that's generally not good practice in a relational database.  I could also give it it's own table.  Thoughts?

**Comment by SteveIreland** [comment]
Sounds like we collectively agree to "object to this".

**Comment by daN4cat** [comment]
Two things.

What if you don't use category and add your own classification attribute?
That could do without changing too much as you have what you need in attributes.
Then build a generic attributes report here.
You could just hide in your branch the category from the input form or make the whole category matter optional in OSPOS.

If not the above...

Where is that we store the attribute options at the moment?
We may use the same table if not puzzling the attributes feature so if one day we move category to attributes all the options are there.

**Comment by objecttothis** [comment]
> What if you don't use category and add your own classification attribute?
> That could do without changing too much as you have what you need in attributes.
> Then build a generic attributes report as that will be handy for everybody.
> You could just hide in your branch the category from the input form or simply make the whole category matter optional in OSPOS.

If at all possible I would like to develop in such a way as to have just one branch of the code.  The reasoning is to avoid the headache of constantly having to maintain changes between two sets of code.  In some ways it would be much easier to write code if I just forked and made all my own design decisions for our company, but that doesn't serve the greater community and unless I keep up on merging changes from opensourcepos I miss out on new features in this repository.  I think it's in everyone's best interest for me to either not make the changes or come up with a way to do it that those making the ultimate design decisions (@daN4cat & @jekkos) are happy.

We all agree that some people want Category to be of TEXT type and others DROPDOWN, so no solution should be considered that doesn't meet those requirements.

With that said, for me the easiest would have been to make Category an attribute and just modify the reports to add the joins.  However, I respect your thoughts and I completely understand your reasoning on why not to do that.  Your suggestion of us just not using Category and instead using our own attribute for the data in Category would be easier than implementing the ability to essentially turn Category into an attribute without making it an attribute (i.e., offer the optional ability to make it a dropdown).  The additional work being that I would need to write an attribute report, but that helps everyone anyway since it would be useful regardless of whether they are disabling Category or not.

If we do it this way we would need to do the following:
- Add a checkbox in the app config -> General view to either "disable Category" or "Hide Category" depending on what wording everyone likes.
- When disabled, Category becomes a non-required field by the item view, gets hidden in the view and links to the Category Summary and Graphical reports get hidden.
- Write a generic Attribute report which would be visible regardless of the status of Category.  Would this report be just identical to the current Category reports only you get to select which Attribute to generate the report on?  I haven't done anything with reports before so I'd have to familiarize myself with how to build the report.  Alternately we could consider the Attribute Report as a different issue/pull request entirely since the report isn't strictly needed for us to "disable Category"

> If not the above...
> 
> Where is that we store the attribute options at the moment?
> We may use the same table if not puzzling the attributes feature so if one day we move category to attributes all the options are there.

Attributes are defined in attribute_definitions.  Adding any rows there will cause problems in the attributes feature because it would show up in the Attributes section when it isn't strictly an attribute.  We could do some acrobatics in the code to make it behave a certain way when it sees an Attribute called Category, but I don't know if anyone would like that option.  It would involve defining it as an Attribute in Attribute_definitions, but not using attribute_links and attribute_values table for referencing the value.  Jury-rigged at best IMO.

My preferences in order of most to least preferred are:
1 - make it an attribute
2 - @daN4cat's first solution above (hiding/disabling Category and using an attribute)
3 - Making the Category optionally a dropdown but leaving it in items.

@daN4cat my understanding of your preferences are:
1 - @daN4cat's first solution above (hiding/disabling Category and using an attribute)
2 - Making the Category optionally a dropdown but leaving it in items.
.
.
.
49 - Eating moldy poop
50 - make it an attribute

;)
Let me know what you want me to do and I'll write the code.  It sounds like your 1/my 2 choice may be the best compromise unless or until making it an attribute climbs in rank above eating moldy poop for you. ;)

**Comment by objecttothis** [comment]
BTW, if we made Category into an attribute, I think we would just get rid of category reports altogether and write Attribute reports instead anyway.

**Comment by jekkos** [comment]
Ok it's present in more reports than just the category specific ones. So it might still be quite some work to change that. 

I also totally understand that you don't want to customize The software and keep your own version in a separate branch. 

But why not just solve it by adding one simple 'category freeze' option in config, that turns the view control into a drop-down and then uses the values available in the table as options? So once you have your set defined you just lock the control and all you can select are all unique values that are present in the database column at that time. Still sounds better than eating any kind of poop.

**Comment by jekkos** [comment]
But on the other hand I think it's also a solution to turn it into an attribute. As you say it won't add extra complexity to the reports if we foresee a generic way of querying for attributes.

But then we need @daN4cat's final agreement.

**Comment by daN4cat** [comment]
My concern is that the attribute story will have a performance hit.
We have a domain modelled database and to generate views and reports I needed to add temp tables to get dimensional views before showing something.
I spent quite some time sorting out that story.

I still haven't bought into this attribute until somebody shows me that it copes well with big databases.
Also the size of the database doubled with the same quantity of data from vs 1.x to 3.x. And when I tried attributes was even worst.

So I ended up adding my three fields across the system and keep merging.
I also considered Category as core data and a taxonomy entity and not an attribute (additional metadata).
Also to keep simple the queries.

**Comment by daN4cat** [comment]
So I find easier to gray out the Category field and build an Attribute report.
Because in any case you need to work on the latter. Also I'm fine with the free text.

**Comment by objecttothis** [comment]
> So I find easier to gray out the Category field and build an Attribute report.

OK, that sounds like the best compromise.  I will add an option in the app config to disable Category which will grey it out in the item view.  If @jekkos is right then I will have to rework a number of reports to omit Category from results but will also hide the Category based reports altogether.  I will also write an attribute report.  One side question is that we talked about moving to a report builder using Reportico.  Would this be a good time to do that or wait until after the Attribute Report is built in our current method of reporting?

**Comment by objecttothis** [comment]
> My concern is that the attribute story will have a performance hit.
> We have a domain modelled database and to generate reports I needed to add temp tables to get dimensional views before showing something.
> I spent quite some time sorting out that story.
> 
> I still haven't bought into this attribute until somebody shows me that it copes well with big databases.
I found several stackexchange questions and other articles all of which said that the additional joins do not affect performance of the database as long as your tables' structure is setup properly (things that are supposed to be indexes are indexes.  My observational experience I see a < 300ms increase in load times of the items table and search results in Items when comparing 3.2.0 and 3.3.0.

> Also the size of the database doubled with the same quantity of data from vs 1.x to 3.x. And when I tried attributes was even worst.
- We have 15,500 rows in our items table. 
- The attribute_values.attribute_value is a Many-to-one relationship so by itself is smaller than < 3.2.1 since in prior versions you had a lot of repeated data in the items table.
- On our database the size doubles between 3.2.0 (20MB) and 3.3.0 (40MB) which is mostly due to attribute_links.  
- Running `OPTIMIZE TABLE ...` on all the tables actually shaved 10MB off the total size of the database for 3.3.0 (50MB -> 40MB) but increased the size of 3.2.0 slightly (19.8MB -> 20.6MB).  Should we add this to the 3.3.0 migration script?
- We have several tables that are empty in the database depending on how the software gets used.  Creating the table only when the feature gets used could save some but less than 1MB.
- Giving the user the option of purging deleted items, people, etc from the database would save some space, but might come at the expense of not being able to see sales history on items that had been deleted previously.

When @jekkos first mentioned the move to Attributes I had a hard time with it because of the same concerns you are expressing.  Now that I see the flexibility it is offering at the cost of negligible performance change it's difficult for me to imagine going back. 

> So I ended up adding my three fields across the system and keep merging.

> Also to keep simple the queries.
- In searching the code for the dbprefix references I found dozens of places in the code where we are using query() instead of Query Builder where Query Builder produces much cleaner looking code.  In most places the Query to select the same data as < 3.2.1 involves added INNER JOINs which in Query Builder equates to one line of code for every JOIN.

**Comment by objecttothis** [comment]
This is now getting into a little bit of a sidebar, but is somewhat related but there are some db structural changes that could reduce overall storage use.  In general our datatypes are a little on the bloated side.  We are using INT where SMALLINT or MEDIUMINT should be used, DECIMAL(15,3) where DECIMAL(6,3) would do, TEXT where VARCHAR might be better, etc.  Here is an example:

**item_quantities**
- Item_id, location_id are both INT(11).  item_id could probably be safely made into a MEDIUMINT() and location_id could probably be safely made into a SMALLINT().  Savings would be 1 and 2 bytes per row respectively. In my database that's 46KB reduction which by itself isn't much but when you consider this this is up to 46KB less data to pull on each query, it can add up in performance gains.
- quantity is DECIMAL(15,3).  Do we really need 15 on the integer side?  What if M was 6 or 7?  On my 3.2.0 database this change alone HALVED the size of item_quantities from 2MB to 1MB.

If database size is really an issue and its impacting performance we could open another issue and put the database on a significant diet by examining where our datatypes are really more than we need them to be.

**Comment by jekkos** [comment]
Ok let's also think of consistency here, if we add foreign keys I feel they should have the same column type as the fields they refer to. Don't feel this is much of an issue for me at this time.

**Comment by objecttothis** [comment]
yes, definitely.  It makes no sense to change item_id to MEDIUMINT() in one table but not in another.  For me the storage size of the database is not a big deal, but it sounds like it is more important to @daN4cat In general performance of the database is important to all of us and if storage size of the data (because the datatype is twice the size that it needs to be) has impact on performance then we should pay attention to it.  

If the performance gains satisfy @daN4cat on attributes then that is less time him merging code changes into his repository and therefore more time spent on other code changes or in things he'd rather be doing.

Unless I'm mistaken, PHP doesn't care if the datatype is an INT or a SMALLINT just as long as it doesn't truncate data.  This means that the migration script is automatic, straight forward and shouldn't require any code changes.

**Comment by jekkos** [comment]
Also regarding the final solution for category, I still feel that using a flag to turn the input field into a drop-down is the easiest way out. Once you have all distinct values entered at least once, the dropdown is just filled with all distinct cases. In that case you only need to add one config option and populate the dropdown with the same query result as we use for suggestions now.

**Comment by daN4cat** [comment]
I'm fine with either.

**Comment by objecttothis** [comment]
@jekkos you're right, that is easier. The only thing about what you are suggesting that isn't clear to me is how an admin is supposed to add a new dropdown option if all we have is a checkbox enabling/disabling the dropdown feature of Category. I suppose an ad hoc solution would be for an admin to uncheck the dropdown category box, then find an item to add the new category type to and then go back and re-check the box.

**Comment by jekkos** [comment]
You're right. But in my case we will not be adding any categories after we have the basic set defined.  Other option could be adding a new category.. item with value -1 and replace the dropdown with a text field if selected.

**Comment by objecttothis** [comment]
Yes, in our case maybe once every 5 years we might add a different category.  I'm just trying to avoid ad hoc solutions if they are going to be long term.  If we think that Category will stay in the items table and not get migrated to Attributes then it would be better to have a solution that provides full functionality.  @daN4cat proposed making it a suggestions but that still allows the user to enter new data.  Your proposal works unless a new Category needs to be added, but has a workaround that isn't too difficult (disable Category as dropdown, add your new Category type in an item then re-enable Category as dropdown).  Alternately storing acceptable Categories for the dropdown in the database either involves creating a new table which in short mimics Attributes or storing the acceptable values for dropdown in a serialized array in the appconfig table which is not best practices, but works.  It also involves more settings in the config (think mimicking the dropdown attribute type functionality in the attribute definition view).  

We have already spent a great deal of time on this discussion so I'll probably just do what you're suggesting and make distinct suggestions the values of the dropdown when "Category as Dropdown" is checked and the user can just uncheck it to add more Category options.

**Comment by dtomlins** [comment]
I'm "gonna" argue the other side - just because.
OS-Commerce and many of the other carts have Attributes that are "properties" of an item, just structured expansions of the description basically.   And sometimes pre-defined.

As well, they also have "Attributes" that are mandatory and required to do a sale.  And may/will change the price.  The "item_id" plus attributeA plus attributeB plus attributeC are REQUIRED to identify.

Think:
   Item:  1  = "Coat".   Yup - all items bundled into 1.  $150.00
   Attrib:  Size:   2,4,6,8,10,12
   Attrib:  Colour:  Red, Blue, Green, Black, Purple
   Attrib:  Hood:  Yes, No       (Yes adds $10.00)
   Attrib:  Lined:  No, Spring (adds $20,00),  Fall (adds $30.00),  Winter (adds $50.00)
So - 240 possible variants, and 7 different possible prices

It "sort-of" works, but it is effectively creating 240 "items" without the entry labour.  Works fine for online ordering, but inventory has to be done for each separately - it IS possible that Black Size 6 sells out first!

Why do "I" have so many "items" in my database?  Well, just like that it exploded!   Almost every product has at least 7 colours ...  a few have 6 sizes and 1,200 colours.  Don't try that with a pull-down list!

Because of the wide "variety", we use an add-on "category_id" field to isolate to the 1,200 categories - again too big for a pull-down.

Yes - there IS a need for a "contained" category grouping list as you mention above, but the "uncontained" still needs to exist for some outside cases.  

Probably the "ideal for most" is to select much like the oft-used "search box" to select one of the existing "categories" as currently used for searching an item or customer.

As long as it's indexed and prefix-only response shouldn't be terrible.  (entered% vs %entered%)

**Comment by josh-tf** [comment]
In our use case we run a few things off the category and wanted to ensure that the right categories were being used (i.e no freeform user input), as we have some other functionality built in based off the category (as well as some external scripts)

Our category is a dropdown:
![image](https://user-images.githubusercontent.com/25597681/59081810-71608880-8933-11e9-91ff-d4caafcedbbb.png)

The list of options is just a csv on our custom options page:

![image](https://user-images.githubusercontent.com/25597681/59081859-af5dac80-8933-11e9-860d-ef9a18d91b5e.png)

Not the ideal solution (csv in wrong format for example will break it), but its how we handled this one

In our actual items (still using custom[n] fields), we have a combination of a dropdown with optional freeform input:

![image](https://user-images.githubusercontent.com/25597681/59081903-dae09700-8933-11e9-9e75-5ed43c4376e6.png)

**Comment by objecttothis** [comment]
@josh-tf Thanks for sharing this.  With the introduction of attributes in 3.3.0, the dropdown functionality (as well as date, decimal and freeform) now exists.  It does not exist however for category due to the discussion about whether to make it an attribute and how to optionally allow dropdown for those that do not want freeform.  I will be picking this back up later once I finish with the API integration that I'm working on now.

**Comment by josh-tf** [comment]
I saw the attribute the other day, looks very useful and a great addition - we have it in backlog to bring the update in to our fork and then we can do away with the custom fields :)

**Comment by objecttothis** [comment]
@daN4cat @jekkos @opensourcepos/org 
I'm getting ready to start the branch for this.  Essentially I will tie this in to the business end of the DROPDOWN Attributes functionality.  There are two ways to implement this.  I need input on which one I should use.

I plan to have a checkbox on Settings->General titled "Make category a dropdown".  That value will be stored in ospos_config as a row.  Then I plan to insert a row into the ospos_attribute_definitions table with 
definition_id -1
definition_name "ospos_category"
definition_type "DROPDOWN"
definition_flags 0

Then in views/items/form.php the logic will be slightly changed to display a form_dropdown if enabled or form_textbox if disabled.

I will have to modify the attribute select logic in the item form to not show category as an addable attribute.

NOW: here is where there are two different ways to approach this that I can see:
Option 1: Implement the adding of the dropdown values in the pre-existing attribute settings area with the rest of the attributes.  I would just need to change the logic in the form.php to prevent it from being converted to any other attribute type.

Option 2: Implement the same logic as in the attribute form.php for adding/removing dropdown values straight into settings->general next to the checkbox to enable the feature.  


The first option is far more eloquent from a UI perspective and less work IMO, but option 2 distinguishes the dropdown from attributes better.  In essense the checkbox is turning Category into a mandatory dropdown attribute while keeping the stored value in the ospos_items table.  

From a database perspective, the value itself is still a part of the core.  From the user perspective it will be the same as a mandatory dropdown attribute (but still in the same place on the page as category is currently, not lumped in with the attributes section).

The thing that I am less a fan about with option 2 is that while I'm still using the mechanism for adding dropdown values from attributes, it doesn't look very good outside of the popup form window and it's more work.  Thoughts please.

**Comment by objecttothis** [comment]
Also, unchecking the box would just mark the definition as deleted so that re-enabling the check restores the dropdown options just like it does for attributes.

**Comment by objecttothis** [comment]
I created draft pull request #2816.

**Comment by objecttothis** [comment]
PR #2816 is under review and this can be closed once that PR is approved and merged.
